baseplate/context/cassandra.py

```diff
@@ -90,7 +96,50 @@ def make_object_for_context(self, name, server_span):
         return cqlmapper.connection.Connection(session_adapter)
 
 
-def _on_execute_complete(_, span):
+class WaitForCallbackResponseFuture(object):
+    """Wrap the ResponseFuture to ensure callbacks have completed.
+
+    This fixes a race condition where the server span can complete before
+    the callback has closed out the child span.
+
+    """
+
+    def __init__(self, span, future, callback_fn, callback_args, errback_fn, errback_args):
+        self.callback_event = Event()
+
+        def register_completion(fn):
+            def register_after_execute(*a, **kw):
+                try:
+                    return fn(*a, **kw)
+                finally:
+                    self.callback_event.set()
+            return register_after_execute
+
+        future.add_callback(register_completion(callback_fn), callback_args)
+        future.add_errback(register_completion(errback_fn), errback_args)
+
+        self.future = future
+
+    def result(self):
+        exc = None
+
+        try:
+            result = self.future.result()
+        except Exception as e:
+            exc = e
+
+        # wait for either _on_execute_complete or _on_execute_failed to run
+        wait_result = self.callback_event.wait(timeout=0.1)
```

Comment:
can we use a (much) smaller wait timeout? we can spend a little more cpu checking in the rare case it really does take that much longer rather than blocking the response for 100ms at minimum if the result won the race.

Reply:
> blocking the response for 100ms at minimum

Isn't this the maximum wait? If `Event.set()` has already been called this will return immediately. If not it will wait/block up to the timeout value.

Reply:
yeah, you're right. ignore me!

Reply:
I'll lower the timeout regardless. I don't think we care to wait around more than ~10ms.